tests/e2e/preferential_topology_utils.go

```diff
@@ -0,0 +1,456 @@
+	framework.Logf("cmd to detach the tag assigned to preferred datastore: %s ", detachTagCat)
+	detachTagCatRes, err := sshExec(sshClientConfig, masterIp, detachTagCat)
+	if err != nil && detachTagCatRes.Code != 0 {
+		fssh.LogResult(detachTagCatRes)
+		return fmt.Errorf("couldn't execute command: %s on host: %v , error: %s",
+			detachTagCat, masterIp, err)
+	}
+	return nil
+}
+
+/*
+	getListOfSharedDatastoresBetweenVMs method is used to fetch the list of datatsores shared between
+
+node vms or shared across entire k8s cluster
+*/
+func getListOfSharedDatastoresBetweenVMs(masterIp string,
+	dataCenter []*object.Datacenter) (map[string]string, error) {
+	var clusFolderTemp []string
+	var clusterFolderName string
+	shareddatastoreListMap := make(map[string]string)
+	for i := 0; i < len(dataCenter); i++ {
+		clusterFolder := govcLoginCmd() + "govc ls " + dataCenter[i].InventoryPath
+		framework.Logf("cmd: %s ", clusterFolder)
+		clusterFolderNameResult, err := sshExec(sshClientConfig, masterIp, clusterFolder)
+		if err != nil && clusterFolderNameResult.Code != 0 {
+			fssh.LogResult(clusterFolderNameResult)
+			return nil, fmt.Errorf("couldn't execute command: %s on host: %v , error: %s",
+				clusterFolder, masterIp, err)
+		}
+		if clusterFolderNameResult.Stdout != "" {
+			clusFolderTemp = strings.Split(clusterFolderNameResult.Stdout, "\n")
+		}
+		for i := 0; i < len(clusFolderTemp)-1; i++ {
+			if strings.Contains(clusFolderTemp[i], "vm") {
+				clusterFolderName = clusFolderTemp[i]
+				break
+			}
+		}
+	}
+	listOfSharedDatastores := govcLoginCmd() +
+		"govc ls " + clusterFolderName + " | xargs -n1 -I% govc object.collect -s % summary.runtime.host | " +
+		"xargs govc datastore.info -H | grep 'Path\\|URL' | tr -s [:space:]"
+	framework.Logf("cmd for fetching list of shared datastores: %s ", listOfSharedDatastores)
+	result, err := sshExec(sshClientConfig, masterIp, listOfSharedDatastores)
+	if err != nil && result.Code != 0 {
+		fssh.LogResult(result)
+		return nil, fmt.Errorf("couldn't execute command: %s on host: %v , error: %s",
+			listOfSharedDatastores, masterIp, err)
+	}
+	sharedDatastoreList := strings.Split(result.Stdout, "\n")
+	for i := 0; i < len(sharedDatastoreList)-1; i = i + 2 {
+		key := strings.ReplaceAll(sharedDatastoreList[i], " Path: ", "")
+		value := strings.ReplaceAll(sharedDatastoreList[i+1], " URL: ", "")
+		shareddatastoreListMap[key] = value
+	}
+	return shareddatastoreListMap, nil
+}
+
+/*
+	getListOfDatastoresByClusterName method is used to fetch the list of datastores accessible to
+
+specific cluster
+*/
+func getListOfDatastoresByClusterName(masterIp string, cluster string) (map[string]string, error) {
+	ClusterdatastoreListMap := make(map[string]string)
+	datastoreListByCluster := govcLoginCmd() +
+		"govc object.collect -s -d ' ' " + cluster + " host | xargs govc datastore.info -H | " +
+		"grep 'Path\\|URL' | tr -s [:space:]"
```

Comment:
I think we can get datastores for each cluster by *object.HostSystem , similar to this method - getHostsByClusterName
Please check if we can use it for preferential.

Reply:
Tried using it in the code but it is not giving the data which I need for verifying datastore preference. Hence used govc way of fetching datastores list.
I need a datastore map which contains both datatsoreName and its respective Url.
For verifying if it is preferred or not, first I am checking with datastoreName and later invoking CNS queryvolume call with datastore url.